server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/sse/S3WithSSECBlobStoreDAO.java

```diff
@@ -0,0 +1,68 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.objectstorage.aws.sse;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.blob.objectstorage.aws.S3BlobStoreConfiguration;
+import org.apache.james.blob.objectstorage.aws.S3BlobStoreDAO;
+import org.apache.james.blob.objectstorage.aws.S3ClientFactory;
+
+import reactor.core.publisher.Mono;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+
+public class S3WithSSECBlobStoreDAO extends S3BlobStoreDAO {
```

Comment:
How do we plan to use `S3WithSSECBlobStoreDAO`? Guice module chooser?

I would prefer to add the SSE-C logic directly into `S3BlobStoreDAO` (we can not have x2 class every time we add an S3 option). WDYT?

Reply:
> How do we plan to use S3WithSSECBlobStoreDAO? Guice module chooser?

It was did in another ticket
https://github.com/apache/james-project/pull/2495

> I would prefer to add the SSE-C logic directly into S3BlobStoreDAO (we can not have x2 class every time we add an S3 option)

I initially considered adding more if else logic to `S3BlobStoreDAO`, but I'd prefer to avoid that (it quiet verbose). While we could implement the interface directly in `S3BlobStoreDAO`, it wouldn't differ much from what `S3WithSSECBlobStoreDAO` does. 
IMO, SSE-C is an option, and it is not simple to inline, -> I prefer to keep the current S3BlobStoreDAO for simple.